### PR TITLE
Move incorrect tx check to the explorer [ECR-3907]

### PR DESCRIPTION
**Pull request by Kormick**
## Overview

Moved incorrect tx check to the explorer. Also added the same check in the websocket handle.

### CODE REVIEW

File: test-suite/testkit/tests/counter/main.rs [hunk 1]
```diff
@@ -1018,3 +1019,35 @@ fn test_explorer_with_before_transactions_error() {
...
+
+    // Send invalid transaction.
+    let (pubkey, key) = crypto::gen_keypair();
+    let tx = Reset.sign(SERVICE_ID + 1, pubkey, &key);
+    let data = hex::encode(tx.to_bytes());
+    let response = api
+        .public(ApiKind::Explorer)
+        .query(&json!({ "tx_body": data }))
+        .post::<TransactionResponse>("v1/transactions")
+        .unwrap_err();
```
Comment (popzxc) [inline test-suite/testkit/tests/counter/main.rs]: Nit: I'd suggest to use `expect_err` and `expect`, since the return value type is the topic of this test. With `expect_err`/`expect` we will receive a more meaningful messages if the test will fail.

File: exonum/src/node/consensus.rs [hunk 2]
```diff
@@ -626,7 +626,7 @@ impl NodeHandler {
             bail!("Received already processed transaction, hash {:?}", hash)
         }
 
-        if let Err(e) = self.blockchain.check_tx(&msg) {
+        if let Err(e) = Blockchain::check_tx(&self.blockchain.snapshot(), &msg) {
```
Comment (popzxc) [inline exonum/src/node/consensus.rs]: Snapshot is already available here, we don't have to create another one. Or am I wrong?

File: exonum/src/blockchain/tests.rs [hunk 3]
```diff
@@ -765,11 +765,11 @@ fn test_check_tx() {
     );
 
     let correct_tx = TestAdd { value: 1 }.sign(TEST_SERVICE_ID, keypair.0, &keypair.1);
-    blockchain.check_tx(&correct_tx).unwrap();
+    Blockchain::check_tx(&blockchain.snapshot(), &correct_tx).unwrap();
```
Comment (popzxc) [inline exonum/src/blockchain/tests.rs]: Nit: you can create a snapshot once to reuse it.

File: exonum/src/runtime/dispatcher/mod.rs [hunk 4]
```diff
@@ -181,11 +181,13 @@ impl Dispatcher {
     /// Returned `Ok(())` value doesn't necessarily mean that transaction is correct and will be
     /// executed successfully, but returned `Err(..)` value means that this transaction is
     /// **obviously** incorrect and should be declined as early as possible.
-    pub(crate) fn check_tx(&self, tx: &Verified<AnyTx>) -> Result<(), ExecutionError> {
-        // Currently the only check is that destination service exists, but later
```
Comment (popzxc) [inline exonum/src/runtime/dispatcher/mod.rs]: Why was the comment removed?

File: exonum/src/api/websocket.rs [hunk 5]
```diff
@@ -346,18 +346,20 @@ impl Handler<Broadcast> for Server {
 impl Handler<Transaction> for Server {
     type Result = Result<TransactionResponse, failure::Error>;
 
+    /// Checks if incoming transaction is valid with Blockchain::check_tx.
+    /// Broadcasts valid transactions and ignores invalid.
```
Comment (popzxc) [inline exonum/src/api/websocket.rs]: Nit: I'd rephrase it like "Broadcasts transaction if the check was passed, and returns an error otherwise.` (something similar can be applied to the `ExplorerApi::add_transaction` as well).

File: exonum/tests/websocket/main.rs [hunk 6]
```diff
@@ -99,6 +100,24 @@ fn test_send_transaction() {
...
+    let tx = CreateWallet::new(pk, "Bob").sign(SERVICE_ID + 1, pk, &sk);
+    let tx_json =
+        serde_json::to_string(&json!({ "type": "transaction", "payload": { "tx_body": tx }}))
+            .unwrap();
+    client.send_message(&OwnedMessage::Text(tx_json)).unwrap();
+
+    // Check response on sent message.
+    let resp_text = recv_text_msg(&mut client).unwrap();
+    assert_eq!(
+        serde_json::from_str::<serde_json::Value>(&resp_text).unwrap(),
```
Comment (slowli) [inline exonum/tests/websocket/main.rs]: Nit: Could make sense to introduce a `let response: Value = ...` binding for this expression for the code to be more readable.

File: exonum/tests/websocket/main.rs [hunk 7]
```diff
@@ -99,6 +100,24 @@ fn test_send_transaction() {
...
+        json!({
+            "result": "error",
+            "description": "Execution error `dispatcher:7` occurred: Suitable runtime for the given service instance ID is not found."
```
Comment (slowli) [inline exonum/tests/websocket/main.rs]: Nit: Overly long line; something like this should work:

```rust
json!({
    "result": "error",
    "description": "Execution error `dispatcher:7` occurred: Suitable runtime \
                    for the given service instance ID is not found.",
})
```

File: exonum/src/blockchain/mod.rs [hunk 8]
```diff
@@ -168,6 +168,15 @@ impl Blockchain {
     pub fn service_keypair(&self) -> &(PublicKey, SecretKey) {
         &self.service_keypair
     }
+
+    /// Performs several shallow checks that transaction is correct.
+    ///
+    /// Returned `Ok(())` value doesn't necessarily mean that transaction is correct and will be
+    /// executed successfully, but returned `Err(..)` value means that this transaction is
+    /// **obviously** incorrect and should be declined as early as possible.
+    pub fn check_tx(snapshot: &dyn Snapshot, tx: &Verified<AnyTx>) -> Result<(), ExecutionError> {
```
Comment (slowli) [inline exonum/src/blockchain/mod.rs]: (Slightly) off-topic: I'm not sure that the API endpoint should return `ExecutionError` as is, especially given the recent work on unifying error format. Could be solved later, I guess.

File: exonum/tests/websocket/main.rs [hunk 9]
```diff
@@ -99,6 +100,24 @@ fn test_send_transaction() {
         })
     );
 
+    // Send invalid transaction.
+    let (pk, sk) = gen_keypair();
+    let tx = CreateWallet::new(pk, "Bob").sign(SERVICE_ID + 1, pk, &sk);
+    let tx_json =
+        serde_json::to_string(&json!({ "type": "transaction", "payload": { "tx_body": tx }}))
```
Comment (slowli) [inline exonum/tests/websocket/main.rs]: Nit: Ugly formatting; binding `json!(..)` to a variable could improve it.

File: exonum/src/blockchain/mod.rs (hunk 8 above)
Comment (popzxc) [reply exonum/src/blockchain/mod.rs]: Yeah, I think we'll handle it within the API errors unification topic.
